src/main/scala/firrtl/passes/LowerTypes.scala

```diff
@@ -100,157 +100,139 @@ object LowerTypes extends Pass {
+          case "addr" | "en" | "clk" | "wmode" =>
+            require(tail.isEmpty) // there can't be a tail for these
+            memDataTypeMap(mem.name) match {
+              case _: GroundType => Seq(e)
+              case memType => create_exps(mem.name, memType) map { e =>
                 val loMemName = loweredName(e)
                 val loMem = WRef(loMemName, UnknownType, kind(mem), UNKNOWNGENDER)
                 mergeRef(loMem, mergeRef(port, field))
               }
-          }
-        // Fields that need not be replicated for each
-        // eg. mem.reader.data[0].a
-        // (Connect/IsInvalid must already have been split to ground types)
-        } else if (Seq("data", "mask", "rdata", "wdata", "wmask").contains(field.name)) {
-          val loMem = tail match {
-            case Some(e) =>
-              val loMemExp = mergeRef(mem, e)
-              val loMemName = loweredName(loMemExp)
-              WRef(loMemName, UnknownType, kind(mem), UNKNOWNGENDER)
-            case None => mem
-          }
-          Seq(mergeRef(loMem, mergeRef(port, field)))
-        } else {
-          error(s"Error! Unhandled memory field ${field.name}")
+            }
+          // Fields that need not be replicated for each
+          // eg. mem.reader.data[0].a
+          // (Connect/IsInvalid must already have been split to ground types)
+          case "data" | "mask" | "rdata" | "wdata" | "wmask" =>
+            val loMem = tail match {
+              case Some(e) =>
+                val loMemExp = mergeRef(mem, e)
+                val loMemName = loweredName(loMemExp)
+                WRef(loMemName, UnknownType, kind(mem), UNKNOWNGENDER)
+              case None => mem
+            }
+            Seq(mergeRef(loMem, mergeRef(port, field)))
+          case name => error(s"Error! Unhandled memory field ${name}")
         }
       }
 
       def lowerTypesExp(e: Expression): Expression = e match {
-        case e: WRef => e
         case (_: WSubField | _: WSubIndex) => kind(e) match {
           case k: InstanceKind =>
             val (root, tail) = splitRef(e)
             val name = loweredName(tail)
             WSubField(root, name, e.tpe, gender(e))
           case k: MemKind =>
             val exps = lowerTypesMemExp(e)
-            if (exps.length > 1)
-              error("Error! lowerTypesExp called on MemKind SubField that needs" +
-                    " to be expanded!")
-            exps(0)
-          case k =>
-            WRef(loweredName(e), e.tpe, kind(e), gender(e))
+            exps.size match {
+              case 1 => exps.head
+              case _ => error("Error! lowerTypesExp called on MemKind " + 
+                              "SubField that needs to be expanded!")
+            }
+          case _ => WRef(loweredName(e), e.tpe, kind(e), gender(e))
         }
         case e: Mux => e map (lowerTypesExp)
         case e: ValidIf => e map (lowerTypesExp)
-        case (_: UIntLiteral | _: SIntLiteral) => e
```

Comment:
Yeah, it will error if a WSubAccess leaks into lower types (a compiler bug). Just make the error check explicit